File: polyglot/python/dmarc_parser_validator.py
"""
dmarcaudit: DMARC/SPF/DKIM Parser Validator

A production-grade parser and validator for email authentication records.
Analyzes domain posture and generates prioritized fix recommendations.

Usage:
    python -m polyglot.python.dmarc_parser_validator <domain> [--verbose]
"""

import argparse
import base64
import binascii
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Optional


class PolicyState(Enum):
    """DMARC policy states."""
    NONE = "none"
    QUARANTINE = "quarantine"
    REJECT = "reject"
    SUBMIT = "submit"  # Legacy, rarely used


class AlignmentMode(Enum):
    """DKIM/SPF alignment modes."""
    RELAXED = "relaxed"
    STRICT = "strict"


@dataclass
class DMARCRecord:
    """Parsed and validated DMARC record data."""
    domain: str
    flags: list[str]
    policy: PolicyState = PolicyState.NONE
    rua: list[str] = field(default_factory=list)
    ruf: list[str] = field(default_factory=list)
    adkim: AlignmentMode = AlignmentMode.RELAXED
    aspf: AlignmentMode = AlignmentMode.RELAXED
    sp: int = 0
    pct: int = 100
    text: str = ""
    raw: dict[str, str] = field(default_factory=dict)
    
    @classmethod
    def from_text(cls, domain: str, text: str) -> "DMARCRecord":
        """Parse a DMARC TXT record."""
        flags = []
        policy = PolicyState.NONE
        rua = []
        ruf = []
        adkim = AlignmentMode.RELAXED
        aspf = AlignmentMode.RELAXED
        sp = 0
        pct = 100
        
        # Normalize and split
        text = text.strip()
        if not text:
            return cls(domain, flags, policy, rua, ruf, adkim, aspf, sp, pct, text)
        
        parts = text.split()
        
        for part in parts:
            if "=" in part:
                key, value = part.split("=", 1)
                key = key.lower().strip("-")
                value = value.strip('"')
                
                # Policy
                if key == "p":
                    policy = PolicyState(value.lower())
                    
                # Reporting addresses
                elif key in ("rua", "ruf"):
                    addr_list = [addr.strip() for addr in value.split(",")]
                    rua if key == "rua" else ruf  # type: ignore
                    
                # Alignment modes
                elif key == "adkim":
                    adkim = AlignmentMode(value.lower())
                    
                elif key == "aspf":
                    aspf = AlignmentMode(value.lower())
                    
                # Subdomain policy
                elif key == "sp":
                    try:
                        sp = int(value)
                    except ValueError:
                        pass
                        
                # Percentage (legacy)
                elif key == "pct":
                    try:
                        pct = int(value)
                    except ValueError:
                        pass
        
        return cls(domain, flags, policy, rua, ruf, adkim, aspf, sp, pct, text)


@dataclass
class SPFRecord:
    """Parsed and validated SPF record data."""
    domain: str
    parts: list[str] = field(default_factory=list)
    auth_rate: float = 100.0
    includes: int = 0
    redirects: int = 0
    has_all: bool = False
    all_type: str = ""  # "-all", "+all", "~all"
    raw: dict[str, str] = field(default_factory=dict)
    
    @classmethod
    def from_text(cls, domain: str, text: str) -> "SPFRecord":
        """Parse an SPF TXT record."""
        parts = []
        includes = 0
        redirects = 0
        has_all = False
        all_type = ""
        
        text = text.strip()
        if not text:
            return cls(domain, parts, 100.0, includes, redirects, has_all, all_type)
        
        # Split by whitespace (SPF uses space as delimiter)
        for token in text.split():
            part = token.strip('"')
            
            # Check for include/redirect
            if part.startswith("include:"):
                parts.append(part)
                includes += 1
                
            elif part.startswith("redirect="):
                parts.append(part)
                redirects += 1
                
            # Check for -all, +all, ~all
            elif part in ("-all", "+all", "~all"):
                has_all = True
                all_type = part
        
        return cls(domain, parts, 100.0, includes, redirects, has_all, all_type)


@dataclass
class DKIMRecord:
    """Parsed and validated DKIM public key record."""
    domain: str
    selector: str
    key: str
    algorithm: str = "rsa-sha256"  # Default assumption
    key_length: int = 0
    is_valid: bool = True
    issues: list[str] = field(default_factory=list)
    
    @classmethod
    def from_text(cls, domain: str, selector: str, text: str) -> "DKIMRecord":
        """Parse a DKIM public key record."""
        key = text.strip()
        
        if not key:
            return cls(domain, selector, "", 0, False, ["Empty key"])
        
        # Extract algorithm and base64 key
        parts = key.split(" ", 2)
        algorithm = "rsa-sha256"
        b64_key = ""
        
        if len(parts) >= 3:
            algorithm = parts[1]
            try:
                b64_key = parts[2].strip()
            except IndexError:
                pass
        
        # Decode and validate base64
        issues = []
        key_length = 0
        is_valid = True
        
        if b64_key:
            try:
                decoded = base64.b64decode(b64_key)
                key_length = len(decoded)
                
                # Check for common issues
                if key_length < 128:
                    issues.append(f"Key too short ({key_length} bytes)")
                    
                elif key_length > 4096:
                    issues.append("Key unusually long")
                    
                # Check algorithm consistency
                expected_len = {
                    "rsa-sha256": 128,
                    "rsa-sha384": 192,
                    "rsa-sha512": 256,
                    "ed25519": 32,
                }.get(algorithm)
                
                if expected_len and key_length != expected_len:
                    issues.append(f"Key length mismatch for {algorithm}")
                    
            except binascii.Error as e:
                is_valid = False
                issues.append(f"Invalid base64 encoding: {e}")
        
        return cls(domain, selector, b64_key, algorithm, key_length, is_valid, issues)


class DMARCAuditResult:
    """Aggregated audit results with spoofability score."""
    
    def __init__(self, domain: str):
        self.domain = domain
        self.dmarc: Optional[DMARCRecord] = None
        self.spf: Optional[SPFRecord] = None
        self.dkims: list[DKIMRecord] = []
        self.timestamp = datetime.utcnow()
        
    def add_dmarc(self, record: DMARCRecord) -> None:
        """Add a DMARC record."""
        if not self.dmarc or (self.dmarc.domain != domain):
            self.dmarc = record
            
    def add_spf(self, record: SPFRecord) -> None:
        """Add an SPF record."""
        if not self.spf or (self.spf.domain != domain):
            self.spf = record
            
    def add_dkim(self, record: DKIMRecord) -> None:
        """Add a DKIM record."""
        if not self.dkims or (self.dkims[-1].selector != record.selector):
            self.dkims.append(record)


def calculate_spoofability_score(result: DMARCAuditResult) -> tuple[float, list[str]]:
    """
    Calculate overall spoofability score (0-100).
    
    Returns:
        Tuple of (score, prioritized_fixes)
    """
    fixes = []
    base_score = 100.0
    
    # DMARC scoring (max -30 points)
    if result.dmarc:
        policy = result.dmarc.policy
        
        if policy == PolicyState.NONE:
            base_score -= 25
            fixes.append(("DMARC", "HIGH", 
                f"Enable DMARC with p=quarantine or p=reject. Current: {policy.value}"))
                
        elif policy == PolicyState.QUARANTINE:
            base_score -= 10
            
        elif policy == PolicyState.REJECT:
            base_score -= 2
            
        # Check alignment modes
        if result.dmarc.adkim == AlignmentMode.RELAXED:
            base_score -= 3
            fixes.append(("DMARC", "MEDIUM", 
                f"Use adkim=strict for tighter security"))
                
        if result.dmarc.aspf == AlignmentMode.RELAXED:
            base_score -= 2
            
        # Check reporting addresses
        if not result.dmarc.rua:
            base_score -= 5
            fixes.append(("DMARC", "MEDIUM", 
                f"Add rua=mailto:<address> for monitoring"))
                
    # SPF scoring (max -10 points)
    if result.spf:
        has_all = result.spf.has_all
        
        if not has_all:
            base_score -= 15
            fixes.append(("SPF", "HIGH", 
                f"Add -all to SPF record. Current parts: {len(result.spf.parts)}"))
                
        elif result.spf.all_type == "+all":
            base_score -= 8
            fixes.append(("SPF", "MEDIUM", 
                f"Change +all to -all for proper failover"))
                
    # DKIM scoring (max -15 points)
    dkim_issues = []
    
    if result.dkims:
        valid_count = sum(1 for d in result.dkims if d.is_valid)
        
        if not any(d.is_valid for d in result.dkims):
            base_score -= 15
            fixes.append(("DKIM", "HIGH", 
                f"Fix or regenerate DKIM keys. {len(result.dkims)} selectors checked."))
                
        elif valid_count < len(result.dkims):
            invalid = [d.selector for d in result.dkims if not d.is_valid]
            base_score -= 10
            fixes.append(("DKIM", "HIGH", 
                f"Fix invalid keys: {', '.join(invalid)}"))
                
        # Check key lengths
        short_keys = [d.selector for d in result.dkims 
                     if d.key_length and d.key_length < 128]
        
        if short_keys:
            base_score -= 5
            fixes.append(("DKIM", "LOW", 
                f"Consider regenerating keys with proper length: {', '.join(short_keys)}"))
                
    # Calculate final score (min 0)
    final_score = max(0.0, base_score)
    
    return round(final_score, 1), fixes


def format_report(result: DMARCAuditResult, 
                  score: float, 
                  fixes: list[tuple[str, str, str]]) -> str:
    """Format the audit report as a readable string."""
    lines = []
    
    # Header
    lines.append("=" * 60)
    lines.append(f"DMARC/SPF/DKIM Audit Report for: {result.domain}")
    lines.append(f"Timestamp: {result.timestamp.strftime('%Y-%m-%d %H:%M:%S UTC')}")
    lines.append("=" * 60)
    
    # Spoofability Score
    lines.append("")
    lines.append(f"SPOOFABILITY SCORE: {score}/100")
    
    if score >= 90:
        status = "GOOD"
    elif score >= 70:
        status = "FAIR"
    elif score >= 50:
        status = "NEEDS ATTENTION"
    else:
        status = "HIGH RISK"
        
    lines.append(f"Status: {status}")
    
    # DMARC Section
    if result.dmarc:
        lines.append("")
        lines.append("-" * 40)
        lines.append("DMARC RECORD")
        lines.append("-" * 40)
        
        policy_str = result.dmarc.policy.value.upper()
        lines.append(f"  Policy (p): {policy_str}")
        lines.append(f"  Subdomain Policy (sp): {result.dmarc.sp}")
        lines.append(f"  Percentage (pct): {result.dmarc.pct}")
        
        if result.dmarc.rua:
            lines.append(f"  Reporting Addresses (rua):")
            for addr in result.dmarc.rua:
                lines.append(f"    - {addr}")
                
        if result.dmarc.ruf:
            lines.append(f"  Forensic Addresses (ruf):")
            for addr in result.dmarc.ruf:
                lines.append(f"    - {addr}")
                
        alignment = f"{result.dmarc.adkim.value.upper()}/{result.dmarc.aspf.value.upper()}"
        lines.append(f"  Alignment: {alignment}")
        
    # SPF Section
    if result.spf:
        lines.append("")
        lines.append("-" * 40)
        lines.append("SPF RECORD")
        lines.append("-" * 40)
        
        has_all_str = f"{result.spf.all_type}" if result.spf.has_all else "(missing -all)"
        lines.append(f"  Has -all: {has_all_str}")
        lines.append(f"  Includes/Redirects: {result.spf.includes}/{result.spf.redirects}")
        
    # DKIM Section
    if result.dkims:
        lines.append("")
        lines.append("-" * 40)
        lines.append("DKIM RECORDS")
        lines.append("-" * 40)
        
        for dkim in result.dkims:
            status = "VALID" if dkim.is_valid else f"INVALID ({dkim.algorithm})"
            lines.append(f"  Selector '{dkim.selector}': {status}")
            
            if dkim.key_length:
                lines.append(f"    Key Length: {dkim.key_length} bytes")
                
            if dkim.issues:
                for issue in dkim.issues:
                    lines.append(f"    Issue: {issue}")
                    
    # Prioritized Fixes
    if fixes:
        lines.append("")
        lines.append("-" * 40)
        lines.append("PRIORITIZED FIXES")
        lines.append("-" * 40)
        
        for priority, category, fix in sorted(fixes, 
                                               key=lambda x: {"HIGH": 0, "MEDIUM": 1, "LOW": 2}.get(x[1], 3)):
            marker = {0: "[HIGH]", 1: "[MEDIUM]", 2: "[LOW]"}.get(priority, "")
            lines.append(f"  {marker} {category}: {fix}")
            
    # Summary
    lines.append("")
    lines.append("=" * 60)
    lines.append("SUMMARY")
    lines.append("=" * 60)
    
    if score >= 90:
        summary = "Domain has strong email authentication posture."
    elif score >= 70:
        summary = "Domain is reasonably secure but has minor improvements needed."
    elif score >= 50:
        summary = "Domain requires attention to reduce spoofability risk."
    else:
        summary = "Domain is vulnerable to email spoofing attacks."
        
    lines.append(f"  {summary}")
    lines.append("")
    
    return "\n".join(lines)


def fetch_dns_records(domain: str, 
                      record_type: str = "TXT",
                      subdomain: Optional[str] = None) -> list[str]:
    """Fetch DNS TXT records for a domain.
    
    Uses standard library dns module if available, otherwise falls back to simple parsing.
    """
    try:
        import dns.resolver
        import dns.rdtypes.IN.TXT
        
        resolver = dns.resolver.Resolver()
        
        # Build query name
        if subdomain:
            query_name = f"{subdomain}.{domain}"
        else:
            query_name = domain
            
        records = []
        for rdata in resolver.resolve(query_name, record_type):
            if isinstance(rdata, dns.rdtypes.IN.TXT.TXT):
                # TXT records can be split across multiple packets
                text_parts = [str(x) for x in rdata.text]
                records.append(" ".join(text_parts))
                
        return records
        
    except ImportError:
        pass
    except Exception as e:
        print(f"DNS

File: polyglot/rust/dmarc_parser_validator.rs
use std::fmt;
use std::str::FromStr;

/// Policy types for DMARC enforcement
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DmarcPolicy {
    /// No action taken (default)
    None,
    /// Put suspicious mail in quarantine
    Quarantine,
    /// Reject suspicious mail
    Reject,
}

impl fmt::Display for DmarcPolicy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DmarcPolicy::None => write!(f, "none"),
            DmarcPolicy::Quarantine => write!(f, "quarantine"),
            DmarcPolicy::Reject => write!(f, "reject"),
        }
    }
}

impl FromStr for DmarcPolicy {
    type Err = ParseDmarcError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let trimmed = s.trim().to_lowercase();
        match trimmed.as_str() {
            "none" => Ok(DmarcPolicy::None),
            "quarantine" | "q" => Ok(DmarcPolicy::Quarantine),
            "reject" | "r" => Ok(DmarcPolicy::Reject),
            "" => Err(ParseDmarcError::EmptyPolicy),
            _ => Err(ParseDmarcError::UnknownPolicy(trimmed)),
        }
    }
}

/// SPF checking mode (ASPF)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AspfMode {
    /// Relaxed: Check if SPF passes, but don't require it
    Relaxed,
    /// Strict: Require SPF to pass
    Strict,
}

impl fmt::Display for AspfMode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AspfMode::Relaxed => write!(f, "r"),
            AspfMode::Strict => write!(f, "s"),
        }
    }
}

impl FromStr for AspfMode {
    type Err = ParseDmarcError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let trimmed = s.trim().to_lowercase();
        match trimmed.as_str() {
            "relaxed" | "r" => Ok(AspfMode::Relaxed),
            "strict" | "s" => Ok(AspfMode::Strict),
            "" => Err(ParseDmarcError::EmptyMode),
            _ => Err(ParseDmarcError::UnknownMode(trimmed)),
        }
    }
}

/// Domain matching mode (ADKIM)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AdkimMode {
    /// Relaxed: Match domain loosely
    Relaxed,
    /// Strict: Exact match required
    Strict,
}

impl fmt::Display for AdkimMode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AdkimMode::Relaxed => write!(f, "r"),
            AdkimMode::Strict => write!(f, "s"),
        }
    }
}

impl FromStr for AdkimMode {
    type Err = ParseDmarcError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let trimmed = s.trim().to_lowercase();
        match trimmed.as_str() {
            "relaxed" | "r" => Ok(AdkimMode::Relaxed),
            "strict" | "s" => Ok(AdkimMode::Strict),
            "" => Err(ParseDmarcError::EmptyMode),
            _ => Err(ParseDmarcError::UnknownMode(trimmed)),
        }
    }
}

/// Parsing errors for DMARC records
#[derive(Debug, Clone)]
pub enum ParseDmarcError {
    EmptyPolicy,
    UnknownPolicy(String),
    EmptyMode,
    UnknownMode(String),
    InvalidPercentage(u16),
    MalformedRecord(String),
}

impl fmt::Display for ParseDmarcError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::EmptyPolicy => write!(f, "Empty policy value"),
            Self::UnknownPolicy(s) => write!(f, "Unknown policy: {}", s),
            Self::EmptyMode => write!(f, "Empty mode value"),
            Self::UnknownMode(s) => write!(f, "Unknown mode: {}", s),
            Self::InvalidPercentage(p) => write!(f, "Invalid percentage: {}", p),
            Self::MalformedRecord(s) => write!(f, "Malformed record: {}", s),
        }
    }
}

impl std::error::Error for ParseDmarcError {}

/// A parsed DMARC record from a TXT value
#[derive(Debug, Clone)]
pub struct DmarcRecord {
    /// The version string (should be "DMARC1")
    pub version: String,
    /// The enforcement policy
    pub policy: DmarcPolicy,
    /// Aggregate report addresses
    pub rua: Vec<String>,
    /// Forensic report addresses (optional)
    pub ruf: Option<Vec<String>>,
    /// Subdomain policy (optional)
    pub sp: Option<DmarcPolicy>,
    /// ADKIM mode
    pub adkim: AdkimMode,
    /// ASPF mode
    pub aspf: AspfMode,
    /// Percentage enforcement (0-100)
    pub pct: u16,
    /// Raw TXT value for debugging
    pub raw: String,
}

impl DmarcRecord {
    /// Parse a DMARC record from a raw TXT string
    /// 
    /// Handles multiple records by returning the first valid one found.
    pub fn parse(txt_value: &str) -> Result<Self, ParseDmarcError> {
        let parts: Vec<&str> = txt_value.split(';').collect();
        
        if parts.is_empty() || parts[0].trim().is_empty() {
            return Err(ParseDmarcError::MalformedRecord(txt_value.to_string()));
        }

        // Parse version (first part)
        let first_part = parts[0].trim();
        let mut version = "DMARC1".to_string();
        
        if !first_part.starts_with("v=") {
            return Err(ParseDmarcError::MalformedRecord(txt_value.to_string()));
        }

        let ver_str = &first_part[2..];
        version = ver_str.trim().to_string();

        // Parse remaining parts
        let mut policy = DmarcPolicy::None;
        let mut rua = Vec::new();
        let mut ruf = None;
        let mut sp = None;
        let mut adkim = AdkimMode::Relaxed;
        let mut aspf = AspfMode::Relaxed;
        let mut pct: u16 = 100;

        for part in &parts[1..] {
            let trimmed = part.trim();
            
            // Skip empty parts
            if trimmed.is_empty() {
                continue;
            }

            // Parse policy (p=)
            if let Some(policy_part) = trimmed.strip_prefix("p=") {
                policy = policy_part.parse()?;
                continue;
            }

            // Parse subdomain policy (sp=)
            if let Some(sp_part) = trimmed.strip_prefix("sp=") {
                sp = Some(sp_part.parse()?);
                continue;
            }

            // Parse aggregate reports (rua=)
            if let Some(rua_part) = trimmed.strip_prefix("rua=") {
                rua.push(rua_part.to_string());
                continue;
            }

            // Parse forensic reports (ruf=)
            if let Some(ruf_part) = trimmed.strip_prefix("ruf=") {
                ruf = Some(vec![ruf_part.to_string()]);
                continue;
            }

            // Parse ADKIM mode (adkim=)
            if let Some(adkim_part) = trimmed.strip_prefix("adkim=") {
                adkim = adkim_part.parse()?;
                continue;
            }

            // Parse ASPF mode (aspf=)
            if let Some(aspf_part) = trimmed.strip_prefix("aspf=") {
                aspf = aspf_part.parse()?;
                continue;
            }

            // Parse percentage (pct=)
            if let Some(pct_part) = trimmed.strip_prefix("pct=") {
                pct = pct_part.parse().map_err(|_| {
                    ParseDmarcError::InvalidPercentage(0)
                })?;
                continue;
            }

            // Unknown parameter - warn but continue
            eprintln!("Warning: unknown DMARC parameter in record: {}", trimmed);
        }

        Ok(DmarcRecord {
            version,
            policy,
            rua,
            ruf,
            sp,
            adkim,
            aspf,
            pct,
            raw: txt_value.to_string(),
        })
    }

    /// Check if this is a valid DMARC1 record
    pub fn is_valid(&self) -> bool {
        self.version == "DMARC1" && !self.rua.is_empty()
    }

    /// Get the effective policy (considering pct=0 disables enforcement)
    pub fn effective_policy(&self) -> DmarcPolicy {
        if self.pct == 0 {
            return DmarcPolicy::None;
        }
        self.policy
    }

    /// Check if reports are configured for aggregate reporting
    pub fn has_aggregate_reports(&self) -> bool {
        !self.rua.is_empty()
    }

    /// Check if forensic reports are configured
    pub fn has_forensic_reports(&self) -> bool {
        self.ruf.as_ref().map_or(false, |v| !v.is_empty())
    }
}

/// SPF record parser and validator
#[derive(Debug, Clone)]
pub struct SpfRecord {
    /// The raw TXT value
    pub raw: String,
    /// Parsed mechanism results
    pub mechanisms: Vec<SpfMechanism>,
    /// Whether SPF is syntactically valid
    pub is_valid: bool,
}

#[derive(Debug, Clone)]
pub enum SpfMechanism {
    /// IP address (IPv4 or IPv6)
    Ip(std::net::IpAddr),
    /// Domain lookup
    Domain(String),
    /// Include another SPF record
    Include(String),
    /// Redirect to another domain
    Redirect(String),
    /// All mail from anywhere (catch-all)
    All,
}

impl SpfRecord {
    pub fn parse(txt_value: &str) -> Self {
        let mechanisms = txt_value.split_whitespace()
            .map(|m| {
                let trimmed = m.trim();
                
                // Handle IP addresses
                if let Ok(ip) = std::net::IpAddr::from_str(trimmed) {
                    return SpfMechanism::Ip(ip);
                }

                // Handle include statements
                if let Some(include_part) = trimmed.strip_prefix("include:") {
                    return SpfMechanism::Include(include_part.to_string());
                }

                // Handle redirect statements
                if let Some(redirect_part) = trimmed.strip_prefix("redirect:") {
                    return SpfMechanism::Redirect(redirect_part.to_string());
                }

                // Handle all
                if trimmed.eq_ignore_ascii_case("-all") || 
                   trimmed.eq_ignore_ascii_case("~all") ||
                   trimmed.eq_ignore_ascii_case("?all") ||
                   trimmed.eq_ignore_ascii_case("+all") {
                    return SpfMechanism::All;
                }

                // Default: treat as domain lookup
                SpfMechanism::Domain(trimmed.to_string())
            })
            .collect();

        // Basic validation - must have at least one mechanism and an all statement
        let has_all = mechanisms.iter().any(|m| matches!(m, SpfMechanism::All));
        
        Self {
            raw: txt_value.to_string(),
            mechanisms,
            is_valid: !mechanisms.is_empty() && has_all,
        }
    }

    pub fn has_strict_policy(&self) -> bool {
        self.mechanisms.last().map_or(false, |m| matches!(m, SpfMechanism::All))
    }

    pub fn is_catch_all(&self) -> bool {
        self.mechanisms.iter()
            .any(|m| matches!(m, SpfMechanism::All))
    }

    pub fn get_include_domains(&self) -> Vec<&str> {
        self.mechanisms.iter()
            .filter_map(|m| match m {
                SpfMechanism::Include(d) => Some(*d),
                _ => None,
            })
            .collect()
    }
}

/// DKIM record parser and validator
#[derive(Debug, Clone)]
pub struct DkimRecord {
    /// The raw TXT value
    pub raw: String,
    /// Extracted domain from selector
    pub domain: Option<String>,
    /// Whether this appears to be a valid DKIM public key
    pub is_valid: bool,
}

impl DkimRecord {
    pub fn parse(txt_value: &str) -> Self {
        let trimmed = txt_value.trim();
        
        // Basic validation - should start with v=DKIM1
        let starts_with_header = trimmed.starts_with("v=DKIM1");
        
        // Try to extract domain from selector
        let domain = if let Some(selector_part) = trimmed.strip_prefix("k=rsa") {
            // Format: k=rsa; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQ...
            // Domain is typically in the header or can be inferred
            if let Some(header_part) = selector_part.strip_prefix("h=") {
                // Header format: h=rsa-sha256; tag=f; s=email.example.com; ...
                if let Some(domain_part) = header_part.strip_prefix("s=") {
                    domain_part.trim().to_string()
                } else {
                    "unknown".to_string()
                }
            } else {
                "unknown".to_string()
            }
        } else {
            "unknown".to_string()
        };

        // Basic validity check
        let has_public_key = trimmed.contains("p=") && 
                           (trimmed.len() > 50 || trimmed.contains("MIG"));

        Self {
            raw: txt_value.to_string(),
            domain: if domain != "unknown" { Some(domain) } else { None },
            is_valid: starts_with_header && has_public_key,
        }
    }

    pub fn get_selector(&self) -> Option<&str> {
        self.raw.split_whitespace()
            .find(|s| s.starts_with("k="))
            .and_then(|s| s.strip_prefix("k="))
    }

    pub fn is_rsa_key(&self) -> bool {
        self.raw.contains("k=rsa")
    }

    pub fn is_ecdsa_key(&self) -> bool {
        self.raw.contains("k=ecdsa-p256") || 
        self.raw.contains("k=ecdsa-p384") ||
        self.raw.contains("k=ecdsa-p521")
    }
}

/// Main audit result with prioritized fixes
#[derive(Debug, Clone)]
pub struct DmarcAuditResult {
    /// The domain being audited
    pub domain: String,
    /// Parsed DMARC record (if found)
    pub dmarc_record: Option<DmarcRecord>,
    /// SPF record analysis
    pub spf_analysis: SpfAnalysis,
    /// DKIM record analysis  
    pub dkim_analysis: DkimAnalysis,
    /// Overall spoofability